radio/enable_cpe_support.py

```diff
@@ -27,6 +27,92 @@ def _forward_cpe(self: VisionTransformer, x: torch.Tensor) -> torch.Tensor:
     return x
 
 
+def _take_indices(
+        num_blocks: int,
+        n: Optional[Union[int, List[int], Tuple[int]]],
+) -> Tuple[Set[int], int]:
+    if isinstance(n, int):
+        assert n >= 0
+        take_indices = {x for x in range(num_blocks - n, num_blocks)}
+    else:
+        take_indices = {num_blocks + idx if idx < 0 else idx for idx in n}
+    return take_indices, max(take_indices)
+
+
+def _forward_intermediates_cpe(
+        self,
+        x: torch.Tensor,
+        indices: Optional[Union[int, List[int], Tuple[int]]] = None,
+        return_prefix_tokens: bool = False,
+        norm: bool = False,
+        stop_early: bool = False,
+        output_fmt: str = 'NCHW',
+        intermediates_only: bool = False,
+        aggregation: Optional[str] = "sparse",
+) -> Union[List[torch.Tensor], Tuple[torch.Tensor, List[torch.Tensor]]]:
+    """ Forward features that returns intermediates.
+    Args:
+        x: Input image tensor
+        indices: Take last n blocks if int, select matching indices if sequence
+        return_prefix_tokens: Return both prefix and spatial intermediate tokens
+        norm: Apply norm layer to all intermediates
+        stop_early: Stop iterating over blocks when last desired intermediate hit
+        output_fmt: Shape of intermediate feature outputs
+        intermediates_only: Only return intermediate features
+        aggregation: intermediate layer aggregation method (sparse or dense)
+    Returns:
+    """
+    assert output_fmt in ('NCHW', 'NLC'), 'Output format must be one of NCHW or NLC.'
+    assert aggregation in ('sparse', 'dense'), 'Aggregation must be one of sparse or dense.'
+    reshape = output_fmt == 'NCHW'
+    intermediates = []
+    take_indices, max_index = _take_indices(len(self.blocks), indices)
+    # forward pass
+    B, _, height, width = x.shape
+    x = self.patch_generator(x)
+
+    if not stop_early:  # can't slice blocks in torchscript
+        blocks = self.blocks
+    else:
+        blocks = self.blocks[:max_index + 1]
+
+    accumulator = 0
+    num_accumulated = 0
+
+    for i, blk in enumerate(blocks):
+        x = blk(x)
+        if aggregation == "dense":
+            accumulator = accumulator + x
+            num_accumulated += 1
+        if i in take_indices:
+            if aggregation == "dense":
+                x_ = accumulator / num_accumulated
+                num_accumulated = 0
+                accumulator = 0
+            else:
+                 x_ = x
+            # normalize intermediates with final norm layer if enabled
+            intermediates.append(self.norm(x_) if norm else x_)
+
+    # process intermediates
+
+    # split prefix (e.g. class, distill) and spatial feature tokens
+    prefix_tokens = [y[:, 0:self.patch_generator.num_cls_tokens] for y in intermediates]
+    intermediates = [y[:, self.patch_generator.num_skip:] for y in intermediates]
+
+    if reshape:
```

Comment:
This is great. Do you think there'd be a way to add this to our regular forward API as well? I know a handful of people have asked how to do this.

Reply:
Yes, we can do something like this to fetch intermediate layers:

```
outputs = model.forward_intermediates(images, indices=[7, 15, 23, 31])
```

I've added an example to the README. Or did I misunderstand your question?

Reply:
I actually meant adding the `reshape` flag to the regular forward call, so that RADIO natively returns the features in BCHW format when `reshape == True`.

Reply:
Ah yes, we can deal with this later!